core/iwasm/fast-jit/fe/jit_emit_memory.c

```diff
@@ -5,18 +5,131 @@
 
 #include "jit_emit_memory.h"
 #include "../jit_frontend.h"
+#include "fe/jit_emit_exception.h"
+
+static JitReg
+check_and_seek(JitCompContext *cc, JitReg addr, JitReg offset, uint32 bytes)
+{
+    JitReg addrL, offset1, memory_boundary, memory_data, maddr;
+    /* the default memory */
+    uint32 mem_idx = 0;
+
+    /* check */
+    /* addrL = (int64_t)addr */
+    addrL = jit_cc_new_reg_I64(cc);
+    GEN_INSN(I32TOI64, addrL, addr);
+
+    /* offset1 = offset + addrL */
+    offset1 = jit_cc_new_reg_I64(cc);
+    GEN_INSN(ADD, offset1, offset, addrL);
+
+    /* if (offset1  >= memory_boundary) goto EXCEPTION */
+    switch (bytes) {
+        case 1:
+        {
+            memory_boundary =
+                get_mem_bound_check_1byte_reg(cc->jit_frame, mem_idx);
+            break;
+        }
+        case 2:
+        {
+            memory_boundary =
+                get_mem_bound_check_2bytes_reg(cc->jit_frame, mem_idx);
+            break;
+        }
+        case 4:
+        {
+            memory_boundary =
+                get_mem_bound_check_4bytes_reg(cc->jit_frame, mem_idx);
+            break;
+        }
+        case 8:
+        {
+            memory_boundary =
+                get_mem_bound_check_8bytes_reg(cc->jit_frame, mem_idx);
+            break;
+        }
+        case 16:
+        {
+            memory_boundary =
+                get_mem_bound_check_16bytes_reg(cc->jit_frame, mem_idx);
+            break;
+        }
+        default:
+        {
+            memory_boundary = get_memory_data_end_reg(cc->jit_frame, mem_idx);
+            break;
+        }
+    }
+
+    GEN_INSN(CMP, cc->cmp_reg, offset1, memory_boundary);
+    if (!jit_emit_exception(cc, EXCE_OUT_OF_BOUNDS_MEMORY_ACCESS, JIT_OP_BGEU,
+                            cc->cmp_reg, 0)) {
+        goto fail;
+    }
+
+    /* seek */
+    memory_data = get_memory_data_reg(cc->jit_frame, mem_idx);
+    maddr = jit_cc_new_reg_I64(cc);
+    GEN_INSN(ADD, maddr, memory_data, offset1);
```

Comment:
Here it is only OK for 64-bit platform, in 32-bit, maddr and memory_data is 32-bit, and offset1 should be converted in 32-bit (I64TOI32) before using it.

In fact in AOT compilation, we don't extend memarg.offset + maddr to uint64, but check (1) whether memarg.offset + maddr is integer overflow, (2) whether the adding result is no larger than memory_boundary. Extending to 64bit, addin, checking and thrinking to 32bit again may introduce a lot instructions and degrade the performance.

Reply:
It seems there are two different comparison groups in AOT. Each for a different platform. Shall we do the same check or use the new solution w/o expanding to i64.

```
    LLVMValueRef offset_const = I32_CONST(offset);
    POP_I32(addr);

    if (is_target_64bit) {
        if (!(offset_const = LLVMBuildZExt(comp_ctx->builder, offset_const,
                                           I64_TYPE, "offset_i64"))
            || !(addr = LLVMBuildZExt(comp_ctx->builder, addr, I64_TYPE,
                                      "addr_i64"))) {
            aot_set_last_error("llvm build zero extend failed.");
            goto fail;
        }
    }

    /* offset1 = offset + addr; */
    BUILD_OP(Add, offset_const, addr, offset1, "offset1");

    if (!(mem_check_bound =
              get_memory_check_bound(comp_ctx, func_ctx, bytes))) {
        goto fail;
    }

    if (is_target_64bit) {
        /* if (offset1 > mem_check_bound) */
        BUILD_ICMP(LLVMIntUGT, offset1, mem_check_bound, cmp, "cmp");
    }
    else {
        /* if (offset1 < addr) */
        /* Check integer overflow */
        BUILD_ICMP(LLVMIntULT, offset1, addr, cmp1, "cmp1");
        /* if (offset1 > mem_check_bound) */
        BUILD_ICMP(LLVMIntUGT, offset1, mem_check_bound, cmp2, "cmp2");
        BUILD_OP(Or, cmp1, cmp2, cmp, "cmp");
    }
```

Reply:
Yes, in AOT compilation, for 64-bit platform, it converts memarg.offset and addr into uint64, add them, and compares the result with mem_inst->mem_bound_check_nbytes:
```C
if ((uint64)memarg.offset + (uint64)addr) > mem_inst->mem_bound_check_nbytes)
  throw out_of_bounds exception;
```
For 32-bit platform:
it checks whether (memarg.offset + addr) is overflow, if not, checks (memarg.offset + addr) > mem_bound_check_nbytes):
```
if ((uint32)memarg.offset + (uint32)addr > (uint32)addr) //integer overflow
   throw out_of_bounds exception;
if ((uint32)memarg.offset + (uint32)addr > mem_bound_check_nbytes)
    throw out_of_bounds exception;
```
Note that mem_bound_check_nbytes is uint64 in 64-bit platform, and uint32 in 32-bit platform.

Seems that it is a little complex? How about we just use the solution for 64-bit platform, and refine it for 32-bit in the future? 64-bit solution is also good for 32-bit platform, but the performance is not the best for 32-bit platform.

Or you just want to unify to use the solution for 32-bit platform? Seems that it is also OK, the performance is also good, not sure whether the performance is better than 64-bit solution.

Reply:
In 64-bit platform, with the solution `if ((uint64)memarg.offset + (uint64)addr) > mem_inst->mem_bound_check_nbytes)`, when checking and getting the final offset, there are 5 instructions:
```bash
extend memarg.offset to 64bit,
extend addr to 64bit
add: offset1 = (uint64)memarg.offset + (uint64)addr)
compare: if (offset1 > mem_bound_check_nbytes)
jxx to throw exception
directly use offset1 to get maddr, as it uint64: maddr = mem_data + offset1
```
with the solution `if ((uint32)memarg.offset + (uint32)addr > (uint32)addr) throw out_of_bounds exception; if ((uint32)memarg.offset + (uint32)addr > mem_bound_check_nbytes) throw out_of_bounds exception;`, when checking and getting the final offset, there are 5 instructions:
```bash
add: offset1 = (uint32)memarg.offset + (uint32)addr
compare, check integer overflow: if (offset1 > (uint32)addr)
jxx to throw exception if integer overflow
compare: if (offset1 > mem_bound_check_nbytes)
jxx to throw exception
extend offset1 to uint64, so maddr = memory_data + (uint64)offset1
```
Not sure whose performance is better, in my memory, the former seems to be better.